src/lib/HST/Util/Selectors.hs

```diff
@@ -89,3 +94,96 @@ getPatVarName (S.PTuple _ _ _) = reportFatal
+getIdentifiersPat :: S.Pat a -> Set String
+getIdentifiersPat (S.PVar _ name)                 = Set.singleton
+  (prettyName name)
+getIdentifiersPat (S.PInfixApp _ pat1 qname pat2) = Set.insert
+  (prettyName qname)
+  (Set.union (getIdentifiersPat pat1) (getIdentifiersPat pat2))
+getIdentifiersPat (S.PApp _ qname pats)           = Set.insert
+  (prettyName qname) (Set.unions (map getIdentifiersPat pats))
+getIdentifiersPat (S.PTuple _ _ pats)             = Set.unions
+  (map getIdentifiersPat pats)
+getIdentifiersPat (S.PParen _ pat)                = getIdentifiersPat pat
+getIdentifiersPat (S.PList _ pats)                = Set.unions
+  (map getIdentifiersPat pats)
+getIdentifiersPat (S.PWildCard _)                 = Set.empty
+
+-- | Collects all identifiers in a right-hand side of a pattern matching rule.
+getIdentifiersRhs :: S.Rhs a -> Set String
+getIdentifiersRhs (S.UnGuardedRhs _ expr) = getIdentifiersExp expr
+getIdentifiersRhs (S.GuardedRhss _ grhss) = Set.unions
+  (map getIdentifiersGRhs grhss)
+
+-- | Collects all identifiers in a guarded right-hand side.
+getIdentifiersGRhs :: S.GuardedRhs a -> Set String
+getIdentifiersGRhs (S.GuardedRhs _ cond expr) = Set.union
+  (getIdentifiersExp cond) (getIdentifiersExp expr)
+
+-- | Collects all identifiers in an expression.
+getIdentifiersExp :: S.Exp a -> Set String
+getIdentifiersExp (S.Var _ qname)            = Set.singleton (prettyName qname)
+getIdentifiersExp (S.Con _ qname)            = Set.singleton (prettyName qname)
+getIdentifiersExp (S.Lit _ _)                = Set.empty
+getIdentifiersExp (S.InfixApp _ exp1 _ exp2) = Set.union
+  (getIdentifiersExp exp1) (getIdentifiersExp exp2)
+getIdentifiersExp (S.App _ exp1 exp2)        = Set.union
+  (getIdentifiersExp exp1) (getIdentifiersExp exp2)
+getIdentifiersExp (S.NegApp _ expr)          = getIdentifiersExp expr
+getIdentifiersExp (S.Lambda _ pats expr)     = Set.union
+  (Set.unions (map getIdentifiersPat pats)) (getIdentifiersExp expr)
+getIdentifiersExp (S.Let _ binds expr)       = Set.union
+  (getIdentifiersBinds binds) (getIdentifiersExp expr)
+getIdentifiersExp (S.If _ cond exp1 exp2)    = Set.union
+  (getIdentifiersExp cond)
+  (Set.union (getIdentifiersExp exp1) (getIdentifiersExp exp2))
+getIdentifiersExp (S.Case _ scrutinee alts)  = Set.union
+  (getIdentifiersExp scrutinee) (Set.unions (map getIdentifiersAlt alts))
+getIdentifiersExp (S.Tuple _ _ exps)         = Set.unions
+  (map getIdentifiersExp exps)
+getIdentifiersExp (S.List _ exps)            = Set.unions
+  (map getIdentifiersExp exps)
+getIdentifiersExp (S.Paren _ expr)           = getIdentifiersExp expr
+getIdentifiersExp (S.ExpTypeSig _ expr _)    = getIdentifiersExp expr
+
+-- | Collects all identifiers in a @case@ alternative.
+getIdentifiersAlt :: S.Alt a -> Set String
+getIdentifiersAlt (S.Alt _ pat rhs binds) = Set.union (getIdentifiersPat pat)
+  (Set.union (getIdentifiersRhs rhs) (maybe Set.empty getIdentifiersBinds binds))
```

Comment:
I'd try to write functions like `Set.union` and `Set.insert` in infix notation to improve readability (i.e., reduce the number of parentheses).

```suggestion
getIdentifiersAlt (S.Alt _ pat rhs binds) =
  getIdentifiersPat pat
    `Set.union` getIdentifiersRhs rhs
    `Set.union` maybe Set.empty getIdentifiersBinds binds
```